File: app/src/main/java/com/sdk/billinglibrary/RemoteConfig.java
package com.sdk.billinglibrary;

import android.content.Context;
import android.util.Log;
import android.util.TypedValue;

import androidx.annotation.Nullable;

import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigValue;

import java.util.HashMap;
import java.util.Map;


public class RemoteConfig {

    static final String KEY_TRIAL = "sub_trial";
    static final String KEY_PREMIUM = "sub_premium";

    static final String KEY_OFFER_WEEKLY = "sub_offer_weekly";
    static final String KEY_OFFER_TRIAL = "sub_offer_trial";
    static final String KEY_OFFER_LIFETIME = "sub_offer_lifetime";

    static final String DEFAULT_OFFER_WEEKLY = "weeklyacessspecialoffer";
    static final String DEFAULT_OFFER_TRIAL = "weeklytrialspecialoffer";
    static final String DEFAULT_OFFER_LIFETIME = "lifetimespecialoffer";

    public interface IOnFetchSubsListener {
        void onComplete(boolean isSuccessful);
    }

    static String getSubByKey(String key) {
        return FirebaseRemoteConfig.getInstance().getString(key);
    }

    static void fetchSubs(Context context, @Nullable IOnFetchSubsListener listener) {
        FirebaseRemoteConfig config = FirebaseRemoteConfig.getInstance();
        Map<String, Object> defaults = new HashMap<>();

        TypedValue trial = new TypedValue();
        TypedValue premium = new TypedValue();

        context.getTheme().resolveAttribute(R.attr.billing_default_premium, premium, true);
        context.getTheme().resolveAttribute(R.attr.billing_default_trial, trial, true);

        defaults.put(KEY_TRIAL, trial.coerceToString());
        defaults.put(KEY_PREMIUM, premium.coerceToString());
        defaults.put(KEY_OFFER_WEEKLY, DEFAULT_OFFER_WEEKLY);
        defaults.put(KEY_OFFER_TRIAL, DEFAULT_OFFER_TRIAL);
        defaults.put(KEY_OFFER_LIFETIME, DEFAULT_OFFER_LIFETIME);

        config.setDefaultsAsync(mergeDefaults(config, defaults)).addOnCompleteListener(t ->
                config.fetchAndActivate().addOnCompleteListener(task -> {

            if (t.isSuccessful())
                Log.d(BillingManager.LOG_TAG, "Defaults are set!");
            else
                Log.d(BillingManager.LOG_TAG, "Defaults are not set!: " + task.getException().toString());

            if (task.isSuccessful())
                Log.d(BillingManager.LOG_TAG, "Fetched new config!");
            else
                Log.d(BillingManager.LOG_TAG, "Fetch failed!");

            String subTrial = config.getString(KEY_TRIAL);
            String subPremium = config.getString(KEY_PREMIUM);
            String offerWeekly = config.getString(KEY_OFFER_WEEKLY);
            String offerTrial = config.getString(KEY_OFFER_TRIAL);
            String offerLifetime = config.getString(KEY_OFFER_LIFETIME);

            Log.d(BillingManager.LOG_TAG, "Trial: " + subTrial);
            Log.d(BillingManager.LOG_TAG, "Premium: " + subPremium);
            Log.d(BillingManager.LOG_TAG, "Offer Weekly: " + offerWeekly);
            Log.d(BillingManager.LOG_TAG, "Offer Trial: " + offerTrial);
            Log.d(BillingManager.LOG_TAG, "Offer Lifetime: " + offerLifetime);

            if (listener != null)
                listener.onComplete(task.isSuccessful());
        }));
    }

    private static Map<String, Object> mergeDefaults(FirebaseRemoteConfig config, Map<String, Object> newDefaults) {
        Map<String, FirebaseRemoteConfigValue> oldValues = config.getAll();
        Map<String, Object> oldDefaults = new HashMap<>();
        for (Map.Entry<String, FirebaseRemoteConfigValue> entry : oldValues.entrySet()) {
            if (entry.getValue().getSource() == FirebaseRemoteConfig.VALUE_SOURCE_DEFAULT)
                oldDefaults.put(entry.getKey(), entry.getValue().asString());
        }
        oldDefaults.putAll(newDefaults);
        return oldDefaults;
    }
}

File: app/src/main/java/com/sdk/billinglibrary/RemoteConfig.kt
package com.sdk.billinglibrary

import android.content.Context
import android.util.Log
import android.util.TypedValue
import com.google.android.gms.tasks.Task
import com.google.firebase.ktx.Firebase
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.firebase.remoteconfig.ktx.remoteConfig

object RemoteConfig {

    const val KEY_TRIAL = "sub_trial"
    const val KEY_PREMIUM = "sub_premium"
    const val KEY_VER = "version"

    fun getVersion(): String {
        return FirebaseRemoteConfig.getInstance().getString(KEY_VER)
    }

    fun getSubByKey(key: String?): String {
        return FirebaseRemoteConfig.getInstance().getString(key!!)
    }

    fun fetchSubs(context: Context, listener: (Boolean) -> Unit) {

        val config = Firebase.remoteConfig
        val defaults = constructDefaults(context)
        
        config.setDefaultsAsync(defaults)
            .addOnCompleteListener { t: Task<Void?> ->
                config.fetchAndActivate().addOnCompleteListener { task: Task<Boolean?> ->

                    if (t.isSuccessful)
                        Log.d(Billing.LOG, "Defaults are set!")
                    else
                        Log.d(Billing.LOG, "Defaults are not set!: ${task.exception.toString()}")

                    if (task.isSuccessful)
                        Log.d(Billing.LOG, "Fetched new config!")
                    else
                        Log.d(Billing.LOG, "Fetch failed!")

                    val subTrial = config.getString(KEY_TRIAL)
                    val subPremium = config.getString(KEY_PREMIUM)
                    val version = config.getString(KEY_VER)

                    Log.d(Billing.LOG, "Trial: $subTrial")
                    Log.d(Billing.LOG, "Premium: $subPremium")
                    Log.d(Billing.LOG, "Version: $version")

                    listener.invoke(task.isSuccessful)
                }
            }
    }

    private fun constructDefaults(context: Context): MutableMap<String, Any> {

        val defaults: MutableMap<String, Any> = HashMap()

        // Fill defaults with already existing defaults (if they were initialized before)
        val oldValues = Firebase.remoteConfig.all
        for ((key, value) in oldValues)
        {
            if (value.source == FirebaseRemoteConfig.VALUE_SOURCE_DEFAULT)
                defaults[key] = value.asString()
        }

        // Add billing defaults
        val trial = TypedValue()
        val premium = TypedValue()
        val version = TypedValue()

        context.theme.resolveAttribute(R.attr.billing_default_premium, premium, true)
        context.theme.resolveAttribute(R.attr.billing_default_trial, trial, true)
        context.theme.resolveAttribute(R.attr.billing_version, version, true)

        defaults[KEY_TRIAL] = trial.coerceToString()
        defaults[KEY_PREMIUM] = premium.coerceToString()
        defaults[KEY_VER] = version.coerceToString()
        
        return defaults
    }
    
}
